ssl/d1_lib.c

```diff
@@ -130,6 +130,17 @@ void dtls1_clear_sent_buffer(SSL_CONNECTION *s)
 
     while ((item = pqueue_pop(s->d1->sent_messages)) != NULL) {
         frag = (hm_fragment *)item->data;
+
+        if (frag->msg_header.is_ccs
+                && frag->msg_header.saved_retransmit_state.wrlmethod != NULL
+                && s->rlayer.wrl != frag->msg_header.saved_retransmit_state.wrl) {
+            /*
+            * If we're freeing the CCS then we're done with the old wrl and it
+            * can bee freed
+            */
+            frag->msg_header.saved_retransmit_state.wrlmethod->free(frag->msg_header.saved_retransmit_state.wrl);
+        }
+
```

Comment:
if we make this change, does it also not need to be mirrored to dtls1_clear_received_buffer, as that path also relies on dtls1_hm_fragment_free doing this check?

Reply:
No. The "w" in wrlmethod/wrl is for "write". The only place that `msg_header.saved_retransmit_state.wrlmethod` and `msg_header.saved_retransmit_state.wrl` are ever set to anything is in `dtls1_buffer_message` which is explicitly about the `sent_messages` queue. So while `dtls1_hm_fragment_free` gets called from `dtls1_clear_received_buffer` it was never doing anything with the `wrl` because it was always NULL in that code path.